target/ppc/translate/vsx-impl.c.inc

```diff
@@ -2124,6 +2124,88 @@ static void gen_xvxsigdp(DisasContext *ctx)
     tcg_temp_free_i64(xbl);
 }
 
+static bool do_lstxv(DisasContext *ctx, arg_D *a, bool store, bool paired)
+{
+    TCGv ea;
+    TCGv_i64 xt;
+    MemOp mop;
+    int offset;
+
+    if (paired) {
+        REQUIRE_INSNS_FLAGS2(ctx, ISA310);
+    }
+
+    REQUIRE_VSX(ctx);
+
+    ea = tcg_temp_new();
+
+    mop = DEF_MEMOP(MO_Q);
+
+    gen_set_access_type(ctx, ACCESS_INT);
+
+    xt = tcg_temp_new_i64();
+
+    if (ctx->le_mode) {
+        do_ea_calc(ctx, a->ra, tcg_constant_tl(a->si + (paired ? 24 : 8)), ea);
+        offset = -8;
+    } else {
+        do_ea_calc(ctx, a->ra, tcg_constant_tl(a->si), ea);
+        offset = 8;
+    }
+
+    if(store) {
+        get_cpu_vsrh(xt, a->rt);
+        tcg_gen_qemu_st_i64(xt, ea, ctx->mem_idx, mop);
+        gen_addr_add(ctx, ea, ea, offset);
+        get_cpu_vsrl(xt, a->rt);
+        tcg_gen_qemu_st_i64(xt, ea, ctx->mem_idx, mop);
+        if (paired) {
+            gen_addr_add(ctx, ea, ea, offset);
+            get_cpu_vsrh(xt, a->rt + 1);
+            tcg_gen_qemu_st_i64(xt, ea, ctx->mem_idx, mop);
+            gen_addr_add(ctx, ea, ea, offset);
+            get_cpu_vsrl(xt, a->rt + 1);
+            tcg_gen_qemu_st_i64(xt, ea, ctx->mem_idx, mop);
+        }
+    } else {
+        tcg_gen_qemu_ld_i64(xt, ea, ctx->mem_idx, mop);
+        set_cpu_vsrh(a->rt, xt);
+        gen_addr_add(ctx, ea, ea, offset);
+        tcg_gen_qemu_ld_i64(xt, ea, ctx->mem_idx, mop);
+        set_cpu_vsrl(a->rt, xt);
+        if (paired) {
+            gen_addr_add(ctx, ea, ea, offset);
+            tcg_gen_qemu_ld_i64(xt, ea, ctx->mem_idx, mop);
+            set_cpu_vsrh(a->rt + 1, xt);
+            gen_addr_add(ctx, ea, ea, offset);
+            tcg_gen_qemu_ld_i64(xt, ea, ctx->mem_idx, mop);
+            set_cpu_vsrl(a->rt + 1, xt);
+        }
+    }
+
+    tcg_temp_free(ea);
+    tcg_temp_free_i64(xt);
+    return true;
+}
+
+static bool do_lstxv_PLS_D(DisasContext *ctx, arg_PLXV *a, bool store, bool paired)
+{
+    arg_D d;
+    REQUIRE_INSNS_FLAGS2(ctx, ISA310);
+    if (unlikely(!resolve_PLS_D(ctx, &d, a))) {
+        return true;
+    }
+
+    return do_lstxv(ctx, &d, store, paired);
+}
+
+TRANS(STXVP, do_lstxv, true, true);
```

Comment:
``REQUIRE_INSNS_FLAGS2(ctx, ISA310)`` is being called inside ``do_lstxv`` when ``paired==true``, and inside ``do_lstxv_PLS_D`` always. This means it's being called on every one of the new instructions, and sometimes twice (pstxvp and plxvp, which are both prefixed and paired).

One way to solve this is to move these checks from the do_* functions to the trans_* functions.

Using the macros I suggested in the original review (and adding ``TRANS64_FLAGS{1,2}``), you could:
```
TRANS_FLAGS2(ISA310, STXVP, do_lstxv, true, true)
TRANS_FLAGS2(ISA310, LXVP, do_lstxv, false, true)
TRANS64_FLAGS2(ISA310, PSTXV, do_lstxv_PLS_D, true, false)
TRANS64_FLAGS2(ISA310, PLXV, do_lstxv_PLS_D, false, false)
TRANS64_FLAGS2(ISA310, PSTXVP, do_lstxv_PLS_D, true, true)
TRANS64_FLAGS2(ISA310, PLXVP, do_lstxv_PLS_D, false, true)
```

Regarding ``TRANS64_FLAGS{1,2}``, I'm not sure we should create it. The same could be accomplished by (for example):
```
TRANS_FLAGS12(64B, ISA310, PSTXV, do_lstxv_PLS_D, true, false)
```
if we *do* define ``TRANS_FLAGS12``, which would be the most generic one. But, since the 64B check is common enough, it might be justifiable to have ``TRANS64_FLAGS*`` too. Opinions?

BTW, drop the ';' after the ``TRANS*`` macros

Reply:
Another way, of course, would be to just call ``REQUIRE_INSNS_FLAGS2(ctx, ISA310)`` inside ``do_lstxv``, without checking for ``paired`` and removing the check from ``do_lstxv_PLS_D``. But I'm still more inclined to use the macros, which *should* be helpful in the long run. And would also do the checks earlier while also being more concise.

Reply:
> Regarding TRANS64_FLAGS{1,2}, I'm not sure we should create it. The same could be accomplished by (for example):

I take that back. REQUIRE_64BIT does more than just checking for the 64B flag. It has a special case to elide code for ppc32.

Reply:
I'm adding ``TRANS_FLAGS``, ``TRANS_FLAGS2``, ``TRANS64_FLAGS`` and ``TRANS64_FLAGS2`` to my DFP series, as I'll also use them there.

For now, I've left out versions that require both ``FLAGS`` and ``FLAGS2`` at the same time. Figured we can add those in the future, if/when needed.